[lldb] Refactor InstrumentationRuntimeAsan and add a new plugin

usama54321 opened the pull request:
[lldb] Refactor InstrumentationRuntimeAsan and add a new plugin InstrumentationRuntimeLibsanitizers.

This commit adds InstrumentationRuntimeLibsanitizers, a new runtime plugin for ASan. The plugin provides the same
functionality as InstrumentationRuntimeASan, but provides a different set of symbols/library names to search for while activating the plugin.

The code can be made a lot cleaner by templatizing the InstrumentationRuntimeASan class, and providing the symbol names through the template argument. However, at the moment the plugin infrastructure does not seem to support templatized plugin classes.

rdar://112491689

usama54321 commented:
@llvm-ci test

JDevlieghere commented on lldb/source/Plugins/InstrumentationRuntime/ASan/CMakeLists.txt:
```diff
@@ -1,5 +1,7 @@
+add_subdirectory(Libsanitizers)
```
Rather than living under the Asan plugin, it seems like his should be living at the same level. If there's shared code between the two plugins, we should hoist that up into a separate library (something like `plugins/Process/Utility`, but for the instrumentation plugins.

usama54321 replied:
I moved the plugin in the parent directory. However, I was not sure how to name the plugin so I have currently named the folder as ASanLibsanitizers and named the files as InstrumentationRuntimeASan.cpp which matches the original ASan plugin naming. I did this mostly to keep the filename/macro short and will appreciate any feedback on this.

usama54321 replied:
More specifically, I have named the file `ASanLibsanitizers/InstrumentationRuntimeASan.cpp` instead of `ASanLibsanitizers/InstrumentationRuntimeASanLibsanitizers.cpp`

JDevlieghere commented on lldb/source/Plugins/InstrumentationRuntime/ASan/Libsanitizers/InstrumentationRuntimeLibsanitizers.cpp:
```diff
@@ -0,0 +1,108 @@
+//===-- InstrumentationRuntimeLibsanitizers.cpp ------------------------------------===//
```
Fix ASCII art

JDevlieghere commented on lldb/source/Plugins/InstrumentationRuntime/ASan/ReportRetriever.h:
```diff
@@ -0,0 +1,31 @@
+//===-- ReportRetriever.h ----------------------------*- C++ -*-===//
```
Fix ASCII art

JDevlieghere commented on lldb/source/Plugins/InstrumentationRuntime/ASan/Libsanitizers/InstrumentationRuntimeLibsanitizers.cpp:
```diff
@@ -0,0 +1,108 @@
+
+void InstrumentationRuntimeLibsanitizers::Deactivate() {
+  if (GetBreakpointID() != LLDB_INVALID_BREAK_ID) {
+    ProcessSP process_sp = GetProcessSP();
+    if (process_sp) {
```
You can do:

```
if (ProcessSP process_sp = GetProcessSP()) {
  ...
}
```

JDevlieghere commented on lldb/source/Plugins/InstrumentationRuntime/ASan/Libsanitizers/InstrumentationRuntimeLibsanitizers.h:
```diff
@@ -0,0 +1,56 @@
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef LLDB_SOURCE_PLUGINS_INSTRUMENTATIONRUNTIME_ASAN_INSTRUMENTATIONRUNTIMELIBSANITIZERS_H
+#define LLDB_SOURCE_PLUGINS_INSTRUMENTATIONRUNTIME_ASAN_INSTRUMENTATIONRUNTIMELIBSANITIZERS_H
```
Moving the Libsanitizers runtime up will also avoid this abomination of a header guard :-)

JDevlieghere commented on lldb/source/Plugins/InstrumentationRuntime/ASan/Libsanitizers/InstrumentationRuntimeLibsanitizers.h:
```diff
@@ -0,0 +1,56 @@
+
+namespace lldb_private {
+
+class InstrumentationRuntimeLibsanitizers : public lldb_private::InstrumentationRuntime {
```
You probably don't need the `lldb_private::` here?

bulbazord commented on lldb/source/Plugins/InstrumentationRuntime/ASan/InstrumentationRuntimeASan.cpp:
```diff
@@ -280,29 +81,13 @@ void InstrumentationRuntimeASan::Activate() {
-
-  if (symbol_address == LLDB_INVALID_ADDRESS)
-    return;
+  bool sync = false;
```
You dropped the `const` qualifier here, does something change `sync` now? If not, please re-add the `const` qualifier.

bulbazord commented on lldb/source/Plugins/InstrumentationRuntime/ASan/Libsanitizers/InstrumentationRuntimeLibsanitizers.cpp:
```diff
@@ -0,0 +1,108 @@
+
+const RegularExpression &
+InstrumentationRuntimeLibsanitizers::GetPatternForRuntimeLibrary() {
+  // FIXME: This shouldn't include the "dylib" suffix.
```
Why shouldn't this include the dylib? The FIXME is pretty clear about what should change but not why. Please add some more detail.

bulbazord commented on lldb/source/Plugins/InstrumentationRuntime/ASan/Libsanitizers/InstrumentationRuntimeLibsanitizers.cpp:
```diff
@@ -0,0 +1,108 @@
+      SetBreakpointID(LLDB_INVALID_BREAK_ID);
+    }
+  }
+  SetActive(false);
```
Does the ordering between the above block and `SetActive(false)` matter? If not, it might be clearer to follow if you move that above the first `if` block. There are a number of nested blocks here.

If you do that, you could also flip the initial condition and return early`GetBreakpointID() == LLDB_INVALID_BREAK_ID`

bulbazord commented on lldb/source/Plugins/InstrumentationRuntime/ASan/ReportRetriever.cpp:
```diff
@@ -0,0 +1,241 @@
+
+  ThreadSP thread_sp =
+      process_sp->GetThreadList().GetExpressionExecutionThread();
+  StackFrameSP frame_sp =
+      thread_sp->GetSelectedFrame(DoNoSelectMostRelevantFrame);
```
Since you're being careful here, you should probably also check the validity of `thread_sp` before using it.

bulbazord commented on lldb/source/Plugins/InstrumentationRuntime/ASan/ReportRetriever.cpp:
```diff
@@ -0,0 +1,241 @@
+
+std::string
+ReportRetriever::FormatDescription(StructuredData::ObjectSP report) {
+  std::string description = std::string(report->GetAsDictionary()
+                                            ->GetValueForKey("description")
+                                            ->GetAsString()
+                                            ->GetValue());
```
`GetAsDictionary()` can fail if we were given a non-dictionary ObjectSP. Might be worth breaking this up and doing some error checking.

bulbazord commented on lldb/source/Plugins/InstrumentationRuntime/ASan/ReportRetriever.cpp:
llvm style is to not use braces for single line if/else/loop blocks

https://llvm.org/docs/CodingStandards.html#don-t-use-braces-on-simple-single-statement-bodies-of-if-else-loop-statements

bulbazord commented on lldb/source/Plugins/InstrumentationRuntime/ASan/ReportRetriever.cpp:
On line 196 you check that `process_sp` is valid before using it, but you use it before checking validity up here. Let's hoist that check higher up.

bulbazord commented on lldb/source/Plugins/InstrumentationRuntime/ASan/ReportRetriever.cpp:
You can combine these in the same way Jonas suggested above:
```
if (auto thread_sp = context->exe_ctx_ref.GetThreadSP())
```
(Use of auto is personal preference here 🙂 )

bulbazord reviewed (commented):
Overall looks alright to me. Aside from the inline feedback, I have 2 general pieces of feedback.
1. It might be easier if you broke this up into 2 separate commits. You're doing 2 things in 1 commit: Refactoring InstrumentationRuntimeAsan and adding a new plugin. That makes it more difficult to review since as I'm reading it, I'm having to determine what each portion of the diff is doing.
2. There were a few places where the formatting looks a bit odd. Could you run clang-format over all the code you changed/added just to make sure? If you haven't done so already, that is.

usama54321 committed:
[lldb] Refactor InstrumentationRuntimeAsan

This commit refactors InstrumentationRuntimeASan by pulling out reusable
code into a separate ReportRetriever class. The purpose of the
refactoring is to allow reuse of the ReportRetriever class in another
plugin.

rdar://112491689

usama54321 committed:
[lldb] Add a new plugin ASanLibsanitizers

This commit adds ASanLibsanitizers, a new runtime plugin for ASan. The
plugin provides the same functionality as InstrumentationRuntimeASan,
but provides a different set of symbols/library names to search for
while activating the plugin.

rdar://112491689

usama54321 commented:
Thanks a lot for the feedback @bulbazord and @JDevlieghere. I have updated the PR. I will appreciate any feedback about the directory/file naming convention after the update.

JDevlieghere commented on lldb/source/Plugins/InstrumentationRuntime/ASanLibsanitizers/InstrumentationRuntimeASan.cpp:
```diff
@@ -0,0 +1,120 @@
+//===-- InstrumentationRuntimeASan.cpp ------------------------------------===//
```
I know it's a bit unwieldy, but should this file be called `InstrumentationRuntimeASanLibsanitizers`?

JDevlieghere reviewed (approved):
LGTM with the filename updated.

bulbazord reviewed (approved):
Thanks for breaking it up, that made it easier for me to review.
I don't see any issues with it right now, so LGTM.

usama54321 commented:
Thanks everyone!
